controllers/openstackcluster_controller.go

```diff
@@ -438,44 +438,75 @@ func reconcileNetworkComponents(log logr.Logger, osProviderClient *gophercloud.P
 			return errors.Errorf("failed to reconcile router: %v", err)
 		}
 	}
-	if !openStackCluster.Spec.ControlPlaneEndpoint.IsValid() {
-		var port int32
-		if openStackCluster.Spec.APIServerPort == 0 {
-			port = 6443
-		} else {
-			port = int32(openStackCluster.Spec.APIServerPort)
-		}
-		fp, err := networkingService.GetOrCreateFloatingIP(openStackCluster, clusterName, openStackCluster.Spec.APIServerFloatingIP)
-		if err != nil {
-			handleUpdateOSCError(openStackCluster, errors.Errorf("Floating IP cannot be got or created: %v", err))
-			return errors.Errorf("Floating IP cannot be got or created: %v", err)
-		}
-		// Set APIEndpoints so the Cluster API Cluster Controller can pull them
-		openStackCluster.Spec.ControlPlaneEndpoint = clusterv1.APIEndpoint{
-			Host: fp.FloatingIP,
-			Port: port,
-		}
-	}
 
 	err = networkingService.ReconcileSecurityGroups(openStackCluster, clusterName)
 	if err != nil {
 		handleUpdateOSCError(openStackCluster, errors.Errorf("failed to reconcile security groups: %v", err))
 		return errors.Errorf("failed to reconcile security groups: %v", err)
 	}
 
+	// Calculate the port that we will use for the API server
+	var apiServerPort int
+	switch {
+	case openStackCluster.Spec.ControlPlaneEndpoint.IsValid():
+		apiServerPort = int(openStackCluster.Spec.ControlPlaneEndpoint.Port)
+	case openStackCluster.Spec.APIServerPort != 0:
+		apiServerPort = openStackCluster.Spec.APIServerPort
+	default:
+		apiServerPort = 6443
+	}
+
 	if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
 		loadBalancerService, err := loadbalancer.NewService(osProviderClient, clientOpts, log)
 		if err != nil {
 			return err
 		}
 
-		err = loadBalancerService.ReconcileLoadBalancer(openStackCluster, clusterName)
+		err = loadBalancerService.ReconcileLoadBalancer(openStackCluster, clusterName, apiServerPort)
 		if err != nil {
 			handleUpdateOSCError(openStackCluster, errors.Errorf("failed to reconcile load balancer: %v", err))
 			return errors.Errorf("failed to reconcile load balancer: %v", err)
 		}
 	}
 
+	if !openStackCluster.Spec.ControlPlaneEndpoint.IsValid() {
+		var host string
+		// If there is a load balancer use the floating IP for it if set, falling back to the internal IP
+		switch {
+		case openStackCluster.Spec.ManagedAPIServerLoadBalancer:
+			if openStackCluster.Status.Network.APIServerLoadBalancer.IP != "" {
```

Comment:
If we need load balancer in this case only, we can move L460-470 in this case.

Reply:
@hidekazuna 

We can't move L460-470 here, because this code only runs if `!openStackCluster.Spec.ControlPlaneEndpoint.IsValid()`.

Consider the case where the user sets `managedAPIServerLoadBalancer: true`, `controlPlaneEndpoint.host: <dns name pointing to floating IP>` and `apiServerFloatingIP: <IP>`. If we moved the code from L460-470 to here, the load-balancer would not be created, even though it would be needed.